### ISSUE TITLE
wpa_supplicant starts before wlan interface is ready

### ISSUE THREAD

**Issue by bennofs** [NixOS/nixpkgs#13136]
Everytime after boot, `wpa_supplicant` doesn't work. The logs give the following hint:

```
-- Reboot --
Feb 19 10:23:22 c-cube systemd[1]: Started WPA Supplicant.
Feb 19 10:23:22 c-cube wpa_supplicant[409]: Successfully initialized wpa_supplicant
Feb 19 10:23:22 c-cube wpa_supplicant[409]: Could not read interface wlp2s0 flags: No such device
Feb 19 10:23:22 c-cube wpa_supplicant[409]: nl80211: Driver does not support authentication/association or connect commands
Feb 19 10:23:22 c-cube wpa_supplicant[409]: nl80211: deinit ifname=wlp2s0 disabled_11b_rates=0
Feb 19 10:23:22 c-cube wpa_supplicant[409]: Could not read interface wlp2s0 flags: No such device
Feb 19 10:23:22 c-cube wpa_supplicant[409]: Could not read interface wlp2s0 flags: No such device
Feb 19 10:23:22 c-cube wpa_supplicant[409]: WEXT: Could not set interface 'wlp2s0' UP
Feb 19 10:23:22 c-cube wpa_supplicant[409]: wlp2s0: Failed to initialize driver interface
Feb 19 10:23:22 c-cube systemd[1]: wpa_supplicant.service: Main process exited, code=exited, status=255/n/a
Feb 19 10:23:22 c-cube systemd[1]: wpa_supplicant.service: Unit entered failed state.
Feb 19 10:23:22 c-cube systemd[1]: wpa_supplicant.service: Failed with result 'exit-code'.
```

It works fine after running `systemctl restart wpa_supplicant` once the system is booted, though. Any ideas on how to fix this?

**Comment by NeQuissimus** [comment]
What versions are you running?
I always have a group of these:

```
Feb 19 22:34:39 nixus systemd[1]: Reached target All Network Interfaces.
Feb 19 22:34:39 nixus systemd[1]: Starting Networking Setup...
Feb 19 22:34:39 nixus systemd[1]: Starting DHCP Client...
Feb 19 22:34:39 nixus wpa_supplicant[642]: Successfully initialized wpa_supplicant
```

However, I am running latest unstable (nixpkgs-16.03pre76763.be0abb3). Have never had issues in the last few months.

**Comment by goblin** [comment]
@bennofs hm, I might have had problems with that in the past, but I'm definitely NOT a maintainer of any nixos module

**Comment by bennofs** [comment]
@goblin You should remove this then: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/wpa_supplicant.nix#L160

**Comment by globin** [comment]
Think you mean me, will have a look later today.

**Comment by copumpkin** [comment]
(amusing that a username mixup somehow ended up with another similarly named user who had experience with NixOS)

**Comment by globin** [comment]
Should be fixed on master.

**Comment by bennofs** [comment]
Thanks, I can confirm that the issue is fixed on master.

**Comment by bennofs** [comment]
@globin looks like it's broken again on nixos-unstable, even with the changes from that commit. 

```
~/nixos (master↑) sudo systemctl show wpa_supplicant | grep network
WantedBy=network.target
Before=shutdown.target network.target
After=sysinit.target network-interfaces.target systemd-journald.socket system.slice basic.target
```

**Comment by NeQuissimus** [comment]
@bennofs Can you look through your `journalctl` ? I just had something odd happening, where `wpa_supplicant` seemed unaware of the interface rename:

```
Mar 10 01:31:37 nixus wpa_supplicant[644]: Could not read interface wlan0 flags: No such device
Mar 10 01:31:37 nixus wpa_supplicant[644]: nl80211: Driver does not support authentication/association or connect commands
Mar 10 01:31:37 nixus wpa_supplicant[644]: nl80211: deinit ifname=wlan0 disabled_11b_rates=0
Mar 10 01:31:37 nixus wpa_supplicant[644]: Could not read interface wlan0 flags: No such device
Mar 10 01:31:37 nixus wpa_supplicant[644]: Could not read interface wlan0 flags: No such device
Mar 10 01:31:37 nixus wpa_supplicant[644]: WEXT: Could not set interface 'wlan0' UP
Mar 10 01:31:37 nixus wpa_supplicant[644]: wlan0: Failed to initialize driver interface
```

I wonder if this is what happened to you.

**Comment by bennofs** [comment]
No I'm still getting:

```
-- Reboot --
Mar 09 15:35:13 c-cube systemd[1]: Started WPA Supplicant.
Mar 09 15:35:13 c-cube wpa_supplicant[705]: Successfully initialized wpa_supplicant
Mar 09 15:35:13 c-cube wpa_supplicant[705]: Could not read interface wlp2s0 flags: No such device
Mar 09 15:35:13 c-cube wpa_supplicant[705]: nl80211: Driver does not support authentication/association or connect commands
Mar 09 15:35:13 c-cube wpa_supplicant[705]: nl80211: deinit ifname=wlp2s0 disabled_11b_rates=0
Mar 09 15:35:13 c-cube wpa_supplicant[705]: Could not read interface wlp2s0 flags: No such device
Mar 09 15:35:13 c-cube wpa_supplicant[705]: Could not read interface wlp2s0 flags: No such device
Mar 09 15:35:13 c-cube wpa_supplicant[705]: WEXT: Could not set interface 'wlp2s0' UP
Mar 09 15:35:13 c-cube wpa_supplicant[705]: wlp2s0: Failed to initialize driver interface
Mar 09 15:35:13 c-cube systemd[1]: wpa_supplicant.service: Main process exited, code=exited, status=255/n/a
Mar 09 15:35:13 c-cube systemd[1]: wpa_supplicant.service: Unit entered failed state.
Mar 09 15:35:13 c-cube systemd[1]: wpa_supplicant.service: Failed with result 'exit-code'.
Mar 09 15:35:49 c-cube systemd[1]: Stopped WPA Supplicant.
Mar 09 15:35:49 c-cube systemd[1]: Started WPA Supplicant.
Mar 09 15:35:49 c-cube wpa_supplicant[1658]: Successfully initialized wpa_supplicant
````

**Comment by NeQuissimus** [comment]
Odd, mine worked fine after a reboot

**Comment by bennofs** [comment]
Fixed in 519673e.
